icicle_v3/backend/cpu/src/curve/cpu_msm.hpp

```diff
@@ -0,0 +1,667 @@
+   * addition execution.
+   * @param base - affine base to be added. Passed as a Per as the input is constant and it saves copying.
+   * @param negate_affine - flag to indicate that the base needs to be subbed instead of added.
+   * @param is_montgomery - flag to indicate that the base is in Montgomery form and first needs to be converted.
+   */
+  void set_phase1_addition_with_affine(const P& bucket, const A base, int bucket_idx)
+  {
+    m_p1 = bucket;
+    m_p2_affine = base;
+    m_return_idx = bucket_idx;
+    m_p2_config = ADD_P1_AND_AFFINE_P1;
+    dispatch();
+  }
+
+  /**
+   * @brief Set up phase 2 addition by value. It is required for the first set of additions - the first line addition
+   * of each segment.
+   * @param line_sum - line sum to be copied because it is also given to triangle sum.
+   * @param bucket - bucket to be added to the line.
+   * @param segment_idx - relevant segment to identify the returning result.
+   */
+  void set_phase2_addition_by_value(const P& line_sum, const P& bucket, int segment_idx)
+  {
+    m_p1 = line_sum;
+    m_p2 = bucket;
+    m_return_idx = segment_idx;
+    m_is_line = true;
+    m_p2_config = ADD_P1_P2_BY_VALUE;
+    dispatch();
+  }
+
+  /**
+   * @brief Set up line addition for phase 2.
+   * @param line_sum - line sum to be copied because it is also given to triangle sum.
+   * @param bucket_ptr - Per to the bucket to be added to the line, passed as a Per to avoid copying.
+   * @param segment_idx - relevant segment to identify the returning result.
+   */
+  void set_phase2_line_addition(const P& line_sum, P* bucket_ptr, const int segment_idx)
+  {
+    m_p1 = line_sum;
+    m_p2_Per = bucket_ptr;
+    m_return_idx = segment_idx;
+    m_is_line = true;
+    m_p2_config = ADD_P1_AND_P2_PER;
+    dispatch();
+  }
+
+  /**
+   * @brief Set up triangle addition for phase 2.
+   * No need to specify segment idx as this is only assigned to a task already handling the triangle's segment idx
+   * (A previous line or triangle sum of this segment).
+   * @param line_sum - line sum to be copied because its value can be updated during the triangle sum execution.
+   * @param triangle_sum_ptr - Per to the triangle sum, passed as a Per to avoid unnecessary copying.
+   */
+  void set_phase2_triangle_addition(P& line_sum, P* triangle_sum_ptr)
+  {
+    m_p1 = line_sum;
+    m_p2_Per = triangle_sum_ptr;
+    m_is_line = false;
+    m_p2_config = ADD_P1_AND_P2_PER;
+    dispatch();
+  }
+
+  /**
+   * @brief Chain addition used in phase 1 when a collision between a new result and an occupied bucket.
+   * @param bucket - the bucket value to be added to the existing result in p1.
+   */
+  void set_phase1_collision_task(P& bucket)
+  {
+    m_p2 = bucket;
+    m_p2_config = ADD_P1_P2_BY_VALUE;
+    dispatch();
+  }
+
+  P m_p1;                 // One of the addends, and holds the addition result afterwards
```

Comment:
addends? is that a word?

Reply:
yes the input element in an addition are called addends